### PR TITLE
pyflyby: init at 1.9.6

### LINKED ISSUE #283206: Package request: pyflyby 1.8.8
**Project description**

Pyflyby is a set of Python programming productivity tools for Python 3.7+.

**Metadata**

* homepage URL: https://deshaw.github.io/pyflyby/
* source URL: https://github.com/deshaw/pyflyby
* license: MIT/X11
* platforms: same as Python

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by jfvillablanca**
## Description of changes
Resolves https://github.com/NixOS/nixpkgs/issues/283206

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/py/pyflyby/package.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  version = "1.8.9";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "deshaw";
+    repo = "pyflyby";
+    rev = version;
+    hash = "sha256-4zDgxq1mSyixnmkSeUXbJMMdg5qhDwZUvcxoOB4jeEI=";
+  };
+
```
Comment (GetPsyched) [inline pkgs/by-name/py/pyflyby/package.nix:17]:
```suggestion

  disabled = python3.pythonOlder "3.7";

```

**Review by GetPsyched — CHANGES_REQUESTED** [review]

**Review by Atry — APPROVED** [review]

File: pkgs/by-name/py/pyflyby/package.nix [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  ];
+
+  pythonImportsCheck = [ "pyflyby" ];
+
+  meta = with lib; {
+    description = "A set of productivity tools for Python";
+    homepage = "https://github.com/deshaw/pyflyby";
+    license = licenses.mit;
+    maintainers = with maintainers; [ jfvillablanca ];
+    mainProgram = "py";
```
Comment (Atry) [inline pkgs/by-name/py/pyflyby/package.nix:39]:
```suggestion
    mainProgram = "tidy-imports";
```

Shouldn't it be `tidy-imports`?

**Review by Atry — CHANGES_REQUESTED** [review]
The `mainProgram` seems not right to me

File: pkgs/by-name/py/pyflyby/package.nix [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "deshaw";
+    repo = "pyflyby";
+    rev = version;
+    hash = "sha256-QkoFr9tFtZ+ZEWlxe9csrzoYFl9/V2l4hKYfUWsXUdc=";
+  };
+  disabled = python3.pythonOlder "3.8";
+
```
Comment (GetPsyched) [inline pkgs/by-name/py/pyflyby/package.nix:19]:
```suggestion
```
The minimum Python version in nixpkgs is 3.9 so this expression wouldn't do much.

**Review by GetPsyched — CHANGES_REQUESTED** [review]

File: pkgs/by-name/py/pyflyby/package.nix (hunk 3 above)
Comment (jfvillablanca) [reply pkgs/by-name/py/pyflyby/package.nix]: I was following upstream's restriction.

Comment (GetPsyched) [reply pkgs/by-name/py/pyflyby/package.nix]: Sorry if I was unclear; I wasn't asking you to not follow upstream's restriction, I simply pointed out that in nixpkgs that restriction is implied in this case (and is even stricter).

File: pkgs/by-name/py/pyflyby/package.nix [hunk 4]
```diff
@@ -0,0 +1,41 @@
+{
+  lib,
+  python3,
+  fetchFromGitHub,
+}:
+
+python3.pkgs.buildPythonApplication rec {
```
Comment (GetPsyched) [inline pkgs/by-name/py/pyflyby/package.nix:7]:
```suggestion
python3.pkgs.buildPythonApplication (finalAttrs: {
```
- https://nix.dev/guides/best-practices#recursive-attribute-set-rec
- #315337

File: pkgs/by-name/py/pyflyby/package.nix [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  pname = "pyflyby";
+  version = "1.9.6";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "deshaw";
+    repo = "pyflyby";
+    rev = version;
```
Comment (GetPsyched) [inline pkgs/by-name/py/pyflyby/package.nix:15]:
```suggestion
    rev = finalAttrs.version;
```

File: pkgs/by-name/py/pyflyby/package.nix [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    ipython
+  ];
+
+  pythonImportsCheck = [ "pyflyby" ];
+
+  meta = with lib; {
+    description = "A set of productivity tools for Python";
+    homepage = "https://github.com/deshaw/pyflyby";
+    license = licenses.mit;
+    maintainers = with maintainers; [ jfvillablanca ];
```
Comment (GetPsyched) [inline pkgs/by-name/py/pyflyby/package.nix:38]:
```suggestion
  meta = {
    description = "A set of productivity tools for Python";
    homepage = "https://github.com/deshaw/pyflyby";
    license = lib.licenses.mit;
    maintainers = with lib.maintainers; [ jfvillablanca ];
```
- https://nix.dev/guides/best-practices#with-scopes
- #208242

File: pkgs/by-name/py/pyflyby/package.nix [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    rev = version;
+    hash = "sha256-QkoFr9tFtZ+ZEWlxe9csrzoYFl9/V2l4hKYfUWsXUdc=";
+  };
+
+  nativeBuildInputs = with python3.pkgs; [
+    setuptools
+    wheel
+  ];
+
+  propagatedBuildInputs = with python3.pkgs; [
```
Comment (GetPsyched) [inline pkgs/by-name/py/pyflyby/package.nix:24]:
```suggestion
  dependencies = with python3.pkgs; [
```

File: pkgs/by-name/py/pyflyby/package.nix [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "deshaw";
+    repo = "pyflyby";
+    rev = version;
+    hash = "sha256-QkoFr9tFtZ+ZEWlxe9csrzoYFl9/V2l4hKYfUWsXUdc=";
+  };
+
+  nativeBuildInputs = with python3.pkgs; [
```
Comment (GetPsyched) [inline pkgs/by-name/py/pyflyby/package.nix:19]:
```suggestion
  build-system = with python3.pkgs; [
```
Not sure if `wheel` would come into this, but feel free to look around

File: pkgs/by-name/py/pyflyby/package.nix [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    setuptools
+    wheel
+  ];
+
+  propagatedBuildInputs = with python3.pkgs; [
+    six
+    toml
+    isort
+    black
```
Comment (GetPsyched) [inline pkgs/by-name/py/pyflyby/package.nix:28]: Do we need black as a runtime dep?

**Review by GetPsyched — CHANGES_REQUESTED** [review]

File: pkgs/by-name/py/pyflyby/package.nix (hunk 9 above)
Comment (jfvillablanca) [reply pkgs/by-name/py/pyflyby/package.nix:28]: It's an odd runtime dependency but build fails without it
```
       > Checking runtime dependencies for pyflyby-1.9.6-py3-none-any.whl
       >   - black not installed
```

File: pkgs/by-name/py/pyflyby/package.nix (hunk 4 above)
Comment (jfvillablanca) [reply pkgs/by-name/py/pyflyby/package.nix:7]: i think fixed-point arguments is currently unsupported for `buildPython*`

https://github.com/NixOS/nixpkgs/pull/271387

Comment (GetPsyched) [reply pkgs/by-name/py/pyflyby/package.nix:7]: In that case, at least use `let in`; `rec` is almost always bad

File: pkgs/by-name/py/pyflyby/package.nix [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  ];
+
+  pythonImportsCheck = [ "pyflyby" ];
+
+  meta = {
+    description = "A set of productivity tools for Python";
+    homepage = "https://github.com/deshaw/pyflyby";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ jfvillablanca ];
+    mainProgram = "tidy-imports";
```
Comment (GetPsyched) [inline pkgs/by-name/py/pyflyby/package.nix:41]: Uh, pretty sure `py` was the correct value for `mainProgram`.
```suggestion
    mainProgram = "py";
```

I'm not sure how many packages would conflict with that binary name, but Python doesn't so IMO it's fine to not rename it.

**Review by GetPsyched — APPROVED** [review]
Derivation LGTM

File: pkgs/by-name/py/pyflyby/package.nix (hunk 2 above)
Comment (Aleksanaa) [reply pkgs/by-name/py/pyflyby/package.nix:41]: This is not solved

File: pkgs/by-name/py/pyflyby/package.nix [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    toml
+    isort
+    black
+    ipython
+  ];
+
+  pythonImportsCheck = [ "pyflyby" ];
+
+  meta = {
+    description = "A set of productivity tools for Python";
```
Comment (Aleksanaa) [inline pkgs/by-name/py/pyflyby/package.nix:37]:
```suggestion
    description = "Set of productivity tools for Python";
```

File: pkgs/by-name/py/pyflyby/package.nix (hunk 2 above)
Comment (GetPsyched) [reply pkgs/by-name/py/pyflyby/package.nix:41]: This was reverted, because why should it be `tidy-imports`? Upstream clearly emphasises `py`.

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**